src/actions/auth.js

```diff
@@ -30,29 +36,77 @@ export const logout = () => (dispatch, getState, { history }) => {
+};
+
 /**
  * Use `hooks/login/useFacebookLogin` as possible
  */
-export const loginWithFB = FB => (dispatch, getState) => {
-  if (FB) {
-    return new Promise(resolve =>
-      FB.login(response => resolve(response), { scope: 'email' }),
-    ).then(response => {
-      if (response.status === authStatus.CONNECTED) {
-        return postAuthFacebookApi({
-          accessToken: response.authResponse.accessToken,
-        })
-          .then(({ token, user: { _id, facebook_id } }) =>
-            dispatch(loginWithToken(token)),
-          )
-          .then(() => authStatus.CONNECTED);
-      } else if (response.status === authStatus.NOT_AUTHORIZED) {
-        dispatch(setLogin(authStatus.NOT_AUTHORIZED));
-      }
-      return response.status;
-    });
+export const loginWithFB = FBSDK => async (dispatch, getState) => {
+  if (!FBSDK) {
+    dispatchLoginErrorToast(dispatch, `[ER0001] ${LOGIN_ERROR_MSG}`);
+    rollbar.error('[ER0001] FB SDK is not ready');
+    throw new Error('[ER0001] FB SDK is not ready');
+  }
+
+  let fbLoginResponse = null;
+  try {
+    // invoke FB SDK Login to get FB-issued access token
+    fbLoginResponse = await FBSDKLogin(FBSDK);
+  } catch (error) {
+    dispatchLoginErrorToast(dispatch, `[ER0002] ${LOGIN_ERROR_MSG}`);
+    rollbar.error(`[ER0002] FB SDK login failed: ${error}`);
+    throw new Error(`[ER0002] FB SDK login failed: ${error}`);
+  }
+
+  if (!fbLoginResponse || !fbLoginResponse.status) {
+    dispatchLoginErrorToast(dispatch, `[ER0003] ${LOGIN_ERROR_MSG}`);
+    rollbar.error(
+      `[ER0003] FB login response is empty or does not have status field`,
+    );
+    throw new Error(
+      `[ER0003] FB login response is empty or does not have status field`,
+    );
+  }
+
+  if (fbLoginResponse.status === authStatus.NOT_AUTHORIZED) {
+    dispatchLoginErrorToast(dispatch, `[ER0004] ${LOGIN_ERROR_MSG}`);
+    rollbar.error(`[ER0004] FB login failed: unauthorized`);
+    dispatch(setLogin(authStatus.NOT_AUTHORIZED));
+    throw new Error(`[ER0004] FB login failed: unauthorized`);
+  }
+
+  if (fbLoginResponse.status === authStatus.CONNECTED) {
+    try {
+      // call GoodJob GraphQL API to get JWT token issued by GoodJob
+      const { token } = await postAuthFacebookApi({
+        accessToken: fbLoginResponse.authResponse.accessToken,
+      });
+      await dispatch(loginWithToken(token));
+    } catch (error) {
+      dispatchLoginErrorToast(dispatch, `[ER0005] ${LOGIN_ERROR_MSG}`);
+      rollbar.error(`[ER0005] Graphql mutation facebookLogin failed: ${error}`);
+      throw new Error(
+        `[ER0005] Graphql mutation facebookLogin failed: ${error}`,
+      );
+    }
+  } else {
+    dispatchLoginErrorToast(dispatch, `[ER0006] ${LOGIN_ERROR_MSG}`);
+    rollbar.error(
+      `[ER0006] FB login failed: unknown auth status: ${fbLoginResponse.status}`,
+    );
+    throw new Error(
+      `[ER0006] FB login failed: unknown auth status: ${fbLoginResponse.status}`,
+    );
```

Comment:
A user may cancel the FB login process, and return to the login modal.

In such case, fbLoginResponse.status may be "canceled"